### PR TITLE
SUMO-226523: Added more validation in the TF

### PR DESCRIPTION
**Pull request by jxie418**
Screenshot 2023-08-22 at 8 46 14 PM
Screenshot 2023-08-22 at 8 45 21 PM
Screenshot 2023-08-22 at 8 47 17 PM

### CODE REVIEW

File: sumologic/resource_sumologic_muting_schedules_library_muting_schedule.go [hunk 1]
```diff
@@ -106,20 +109,26 @@ func getScheduleDefinitionSchemma() map[string]*schema.Schema {
...
-			Type:         schema.TypeString,
-			Required:     true,
-			ForceNew:     false,
-			ValidateFunc: validation.StringLenBetween(5, 5),
+			Type:     schema.TypeString,
+			Required: true,
+			ForceNew: false,
+			ValidateFunc: validation.All(
+				validation.StringLenBetween(5, 5),
+				StringIsValidStartTime(),
```
Comment (rmamilla-sumo) [inline sumologic/resource_sumologic_muting_schedules_library_muting_schedule.go]: wouldn't a regex validation like what we did for schedule (line 113) not work here?

File: sumologic/resource_sumologic_muting_schedules_library_muting_schedule.go [hunk 2]
```diff
@@ -106,20 +109,26 @@ func getScheduleDefinitionSchemma() map[string]*schema.Schema {
 			Type:     schema.TypeString,
 			ForceNew: false,
 			Required: true,
-			ValidateFunc: validation.StringMatch(regexp.MustCompile(`^\d{4}-\d{2}-\d{2}$`),
-				"start date in format of yyyy-mm-dd"),
+			ValidateFunc: validation.All(
+				validation.StringMatch(regexp.MustCompile(`^\d{4}-\d{2}-\d{2}$`),
+					"start date in format of yyyy-mm-dd"),
+				StringIsValidStartDate(),
```
Comment (rmamilla-sumo) [inline sumologic/resource_sumologic_muting_schedules_library_muting_schedule.go]: Curious why this is needed?

File: sumologic/resource_sumologic_muting_schedules_library_muting_schedule.go [hunk 3]
```diff
@@ -301,6 +310,51 @@ func getScheduleDefinition(d *schema.ResourceData) ScheduleDefinition {
...
+		yesterday := time.Now().AddDate(0, 0, -1).Truncate(24 * time.Hour)
+
+		if date.Before(yesterday) {
+			return warnings, []error{fmt.Errorf("expected %q to be an valid start date : got %v", k, v)}
+		}
+		return warnings, errors
+	}
+}
+
+var (
```
Comment (rmamilla-sumo) [inline sumologic/resource_sumologic_muting_schedules_library_muting_schedule.go]: curious why ( after var ?

File: sumologic/resource_sumologic_muting_schedules_library_muting_schedule.go [hunk 4]
```diff
@@ -106,20 +109,26 @@ func getScheduleDefinitionSchemma() map[string]*schema.Schema {
...
+				validation.StringLenBetween(5, 5),
+				StringIsValidStartTime(),
+			),
 		},
 		"duration": {
 			Type:         schema.TypeInt,
 			Required:     true,
 			ForceNew:     false,
-			ValidateFunc: validation.IntAtLeast(0),
+			ValidateFunc: validation.IntAtLeast(15),
```
Comment (rmamilla-sumo) [inline sumologic/resource_sumologic_muting_schedules_library_muting_schedule.go:131]: did we define any limit on the duration? 
does this allow negative value ?

File: sumologic/resource_sumologic_muting_schedules_library_muting_schedule.go (hunk 2 above)
Comment (jxie418) [reply sumologic/resource_sumologic_muting_schedules_library_muting_schedule.go]: The start date should be greater than the current date - 24 hours(buffer time). We do not allow the user to create a muting scheduler that starts past.

File: sumologic/resource_sumologic_muting_schedules_library_muting_schedule.go (hunk 1 above)
Comment (jxie418) [reply sumologic/resource_sumologic_muting_schedules_library_muting_schedule.go]: Updated using regex to validate it.

File: sumologic/resource_sumologic_muting_schedules_library_muting_schedule.go (hunk 3 above)
Comment (jxie418) [reply sumologic/resource_sumologic_muting_schedules_library_muting_schedule.go]: removed it.

File: sumologic/resource_sumologic_muting_schedules_library_muting_schedule.go (hunk 4 above)
Comment (jxie418) [reply sumologic/resource_sumologic_muting_schedules_library_muting_schedule.go:131]: yes. we have a minimum duration of 15 minutes. 
no, it does not allow negative value.

File: sumologic/resource_sumologic_muting_schedules_library_muting_schedule.go [hunk 5]
```diff
@@ -301,6 +310,33 @@ func getScheduleDefinition(d *schema.ResourceData) ScheduleDefinition {
 	return scheduleDefinition
 }
 
+func StringIsValidStartDate() schema.SchemaValidateFunc {
+	return func(i interface{}, k string) (warnings []string, errors []error) {
```
Comment (rmamilla-sumo) [inline sumologic/resource_sumologic_muting_schedules_library_muting_schedule.go:314]: curious what are i and k fields mean here?

File: sumologic/resource_sumologic_muting_schedules_library_muting_schedule.go [hunk 6]
```diff
@@ -301,6 +310,33 @@ func getScheduleDefinition(d *schema.ResourceData) ScheduleDefinition {
 	return scheduleDefinition
 }
 
+func StringIsValidStartDate() schema.SchemaValidateFunc {
```
Comment (rmamilla-sumo) [inline sumologic/resource_sumologic_muting_schedules_library_muting_schedule.go]: if im not wrong the validation of regex ensures if it is following the date format. this is doing -24 hr validation, do u think renaming the method would represent this correctly

File: sumologic/resource_sumologic_muting_schedules_library_muting_schedule.go [hunk 7]
```diff
@@ -59,6 +61,7 @@ func getMutingScheduleBaseSchema() map[string]*schema.Schema {
 			Elem: &schema.Resource{
 				Schema: getMonitorScopeSchema(),
 			},
+			AtLeastOneOf: monitorAtleastOneKey,
```
Comment (rmamilla-sumo) [inline sumologic/resource_sumologic_muting_schedules_library_muting_schedule.go:64]: can we add test cases for this?

File: sumologic/resource_sumologic_muting_schedules_library_muting_schedule.go (hunk 5 above)
Comment (jxie418) [reply sumologic/resource_sumologic_muting_schedules_library_muting_schedule.go:314]: This is the format for creating a customer SchemaValidationFun

File: sumologic/resource_sumologic_muting_schedules_library_muting_schedule.go (hunk 6 above)
Comment (jxie418) [reply sumologic/resource_sumologic_muting_schedules_library_muting_schedule.go]: updated.

File: sumologic/resource_sumologic_muting_schedules_library_muting_schedule.go (hunk 7 above)
Comment (jxie418) [reply sumologic/resource_sumologic_muting_schedules_library_muting_schedule.go:64]: Sure.

Comment (jxie418) [reply sumologic/resource_sumologic_muting_schedules_library_muting_schedule.go:64]: updated

**Review by rmamilla-sumo — APPROVED** [review]
LGTM

**Pull request merged**